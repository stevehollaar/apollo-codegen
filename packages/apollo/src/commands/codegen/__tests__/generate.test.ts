jest.mock("apollo-codegen-core/lib/localfs", () => {
  return require("../../../__mocks__/localfs");
});

// this is because of herkou-cli-utils hacky mocking system on their console logger
import { stdout, mockConsole } from "heroku-cli-util";
import * as path from "path";
import * as fs from "fs";
import { test as setup } from "apollo-cli-test";
import { introspectionQuery, print, execute, buildSchema } from "graphql";
import gql from "graphql-tag";
import { fs as mockFS, vol } from "apollo-codegen-core/lib/localfs";

const test = setup.do(() => mockConsole());

const graphQLSchema = fs.readFileSync(
  path.resolve(__dirname, "../../schema/__tests__/fixtures/schema.graphql"),
  {
    encoding: "utf-8"
  }
);

const fullSchema = execute(buildSchema(graphQLSchema), gql(introspectionQuery))
  .data;

const simpleQuery = fs.readFileSync(
  path.resolve(__dirname, "./fixtures/simpleQuery.graphql")
);

const otherQuery = fs.readFileSync(
  path.resolve(__dirname, "./fixtures/otherQuery.graphql")
);

const clientSideSchema = fs.readFileSync(
  path.resolve(__dirname, "./fixtures/clientSideSchema.graphql")
);

const serverSideSchemaTag = `
gql\`
type Query {
  hello: String!
  serverSideField: ServerField!
}

type ServerField {
  serverData: String!
}

type RemovedField {
  id: ID!
  name: RemovedType
}

type RemovedType {
  fieldName: String
}
\`
`;

const clientSideSchemaTag = `
gql\`
${clientSideSchema}
\`
`;

const clientSideSchemaQuery = fs.readFileSync(
  path.resolve(__dirname, "./fixtures/clientSideSchemaQuery.graphql")
);

const clientSideOnlySchema = fs.readFileSync(
  path.resolve(__dirname, "./fixtures/clientSideOnlySchema.graphql")
);

const clientSideOnlyQuery = fs.readFileSync(
  path.resolve(__dirname, "./fixtures/clientSideOnlyQuery.graphql")
);

beforeEach(() => {
  vol.reset();
  vol.fromJSON({
    __blankFileSoDirectoryExists: ""
  });
});

jest.setTimeout(25000);

describe("successful codegen", () => {
  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "API.swift"])
    .it("infers Swift target and writes types", () => {
      expect(mockFS.readFileSync("API.swift").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.graphql": graphQLSchema,
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.graphql", "API.swift"])
    .it(
      "infers Swift target and writes types when schema is a GraphQL file",
      () => {
        expect(mockFS.readFileSync("API.swift").toString()).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.js": serverSideSchemaTag,
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.js", "API.swift"])
    .it("infers Swift target and writes types when schema is a JS file", () => {
      expect(mockFS.readFileSync("API.swift").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.ts": serverSideSchemaTag,
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.ts", "API.swift"])
    .it("infers Swift target and writes types when schema is a TS file", () => {
      expect(mockFS.readFileSync("API.swift").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--operationIdsPath=myOperationIDs.json",
      "API.swift"
    ])
    .it("generates operation IDs files when flag is set", () => {
      expect(
        mockFS.readFileSync("myOperationIDs.json").toString()
      ).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString(),
        "queryTwo.graphql": otherQuery.toString(),
        "outDirectory/__create_this_directory": ""
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--only=queryTwo.graphql",
      "--target=swift",
      "outDirectory"
    ])
    .it("handles only flag for Swift target", () => {
      expect(
        Object.entries(vol.toJSON("outDirectory")).map(arr => [
          path.relative(__dirname, arr[0]),
          arr[1]
        ])
      ).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "API.scala"])
    .it("infers Scala target and writes types", () => {
      expect(mockFS.readFileSync("API.scala").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--outputFlat",
      "API.ts"
    ])
    .it("infers TypeScript target and writes types", () => {
      expect(mockFS.readFileSync("API.ts").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "clientSideSchema.graphql": clientSideSchema.toString(),
        "clientSideSchemaQuery.graphql": clientSideSchemaQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--clientSchema=clientSideSchema.graphql",
      "--outputFlat",
      "API.ts"
    ])
    .it(
      "infers TypeScript target and writes types for query with client-side data",
      () => {
        expect(mockFS.readFileSync("API.ts").toString()).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "clientSideSchemaTag.js": clientSideSchemaTag.toString(),
        "clientSideSchemaQuery.graphql": clientSideSchemaQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--clientSchema=clientSideSchemaTag.js",
      "--outputFlat",
      "API.ts"
    ])
    .it(
      "infers TypeScript target and writes types for query with client-side data with schema in a JS file",
      () => {
        expect(mockFS.readFileSync("API.ts").toString()).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "clientSideSchemaTag.js": clientSideSchemaTag.toString(),
        "clientSideSchemaQuery.graphql": clientSideSchemaQuery.toString(),
        "package.json": `
        {
          "apollo": {
            "schemas": {
              "serverSchema": {
                "schema": "schema.json"
              },
              "default": {
                "schema": "clientSideSchemaTag.js",
                "extends": "serverSchema",
                "clientSide": true
              }
            }
          }
        }
        `
      });
    })
    .command(["codegen:generate", "--outputFlat", "API.ts"])
    .it(
      "infers TypeScript target and writes types for query with client-side data with schema in a JS file from config",
      () => {
        expect(mockFS.readFileSync("API.ts").toString()).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "clientSideOnlySchema.graphql": clientSideOnlySchema.toString(),
        "clientSideOnlyQuery.graphql": clientSideOnlyQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--clientSchema=clientSideOnlySchema.graphql",
      "--outputFlat",
      "API.ts"
    ])
    .it(
      "infers TypeScript target and writes types for query with only client-side data",
      () => {
        expect(mockFS.readFileSync("API.ts").toString()).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--outputFlat",
      "API.js"
    ])
    .it("infers Flow target and writes types", () => {
      expect(mockFS.readFileSync("API.js").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--useFlowExactObjects",
      "--outputFlat",
      "API.js"
    ])
    .it("writes exact Flow types when the flag is set", () => {
      expect(mockFS.readFileSync("API.js").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--useFlowReadOnlyTypes",
      "--outputFlat",
      "API.js"
    ])
    .it("writes read-only Flow types when the flag is set", () => {
      expect(mockFS.readFileSync("API.js").toString()).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "queryOne.graphql": simpleQuery.toString()
      });
    })
    .command(["codegen:generate", "--schema=schema.json", "operations.json"])
    .it("infers JSON target and writes operations", () => {
      expect(
        mockFS.readFileSync("operations.json").toString()
      ).toMatchSnapshot();
    });

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.tsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.tsx",
      "--target=typescript"
    ])
    .it(
      "writes TypeScript types into a __generated__ directory next to sources when no output is set",
      () => {
        expect(
          mockFS
            .readFileSync("directory/__generated__/SimpleQuery.ts")
            .toString()
        ).toMatchSnapshot();
        expect(
          mockFS.readFileSync("__generated__/globalTypes.ts").toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.jsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.jsx",
      "--target=flow"
    ])
    .it(
      "writes Flow types into a __generated__ directory next to sources when no output is set",
      () => {
        expect(
          mockFS
            .readFileSync("directory/__generated__/SimpleQuery.js")
            .toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.tsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.tsx",
      "--target=typescript",
      "__foo__"
    ])
    .it(
      "writes TypeScript types to a custom directory next to sources when output is set",
      () => {
        expect(
          mockFS.readFileSync("directory/__foo__/SimpleQuery.ts").toString()
        ).toMatchSnapshot();
        expect(
          mockFS.readFileSync("__foo__/globalTypes.ts").toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.tsx": `
          gql\`
            query SimpleQuery {
              someEnum
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.tsx",
      "--target=typescript",
      "--globalTypesFile=__foo__/bar.ts"
    ])
    .it(
      "writes TypeScript global types to a custom path when globalTypesFile is set",
      () => {
        expect(
          mockFS
            .readFileSync("directory/__generated__/SimpleQuery.ts")
            .toString()
        ).toMatchSnapshot();
        expect(
          mockFS.readFileSync("__foo__/bar.ts").toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.tsx": `
          gql\`
            query SimpleQuery {
              hello
              someEnum
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.tsx",
      "--target=typescript",
      "--globalTypesFile=__foo__/bar.flow.js"
    ])
    .it(
      "writes Flow global types to a custom path when globalTypesFile is set",
      () => {
        expect(
          mockFS
            .readFileSync("directory/__generated__/SimpleQuery.ts")
            .toString()
        ).toMatchSnapshot();
        expect(
          mockFS.readFileSync("__foo__/bar.flow.js").toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.jsx": `
          gql\`
            query SimpleQuery {
              hello
            }
          \`;
        `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.jsx",
      "--target=flow",
      "__foo__"
    ])
    .it(
      "writes Flow types to a custom directory next to sources when output is set",
      () => {
        expect(
          mockFS.readFileSync("directory/__foo__/SimpleQuery.js").toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.tsx": `
            gql\`
              query SimpleQuery {
                hello
              }
            \`;
          `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.tsx",
      "--target=typescript",
      ""
    ])
    .it(
      "writes TypeScript types next to sources when output is set to empty string",
      () => {
        expect(
          mockFS.readFileSync("directory/SimpleQuery.ts").toString()
        ).toMatchSnapshot();
        expect(
          mockFS.readFileSync("globalTypes.ts").toString()
        ).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      vol.fromJSON({
        "schema.json": JSON.stringify(fullSchema.__schema),
        "directory/component.jsx": `
            gql\`
              query SimpleQuery {
                hello
              }
            \`;
          `
      });
    })
    .command([
      "codegen:generate",
      "--schema=schema.json",
      "--queries=**/*.jsx",
      "--target=flow",
      ""
    ])
    .it(
      "writes Flow types next to sources when output is set to empty string",
      () => {
        expect(
          mockFS.readFileSync("directory/SimpleQuery.js").toString()
        ).toMatchSnapshot();
      }
    );
});

describe("error handling", () => {
  test
    .command(["codegen:generate", "--target=foobar"])
    .catch(err => expect(err.message).toMatch(/Unsupported target: foobar/))
    .it("errors with an unsupported target");

  test
    .command(["codegen:generate", "--target=swift"])
    .catch(err =>
      expect(err.message).toMatch(/The output path must be specified/)
    )
    .it("errors when no output file is provided");

  test
    .command(["codegen:generate", "output-file"])
    .catch(err =>
      expect(err.message).toMatch(
        /Could not infer target from output file type, please use --target/
      )
    )
    .it("errors when target cannot be inferred");
});
