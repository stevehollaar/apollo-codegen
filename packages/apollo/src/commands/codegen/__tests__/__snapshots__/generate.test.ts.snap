// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful codegen generates operation IDs files when flag is set 1`] = `
"{
  \\"4de1e386589b0853a102a87a3f21ca25266116517e59c1e8be9442e2571f00bc\\": {
    \\"name\\": \\"SimpleQuery\\",
    \\"source\\": \\"query SimpleQuery {\\\\n  hello\\\\n}\\"
  }
}"
`;

exports[`successful codegen handles only flag for Swift target 1`] = `
Array [
  Array [
    "../../../../../../outDirectory/__create_this_directory",
    "",
  ],
  Array [
    "../../../../../../outDirectory/Types.graphql.swift",
    "//  This file was automatically generated and should not be edited.

import Apollo",
  ],
  Array [
    "../../../../../../outDirectory/queryTwo.graphql.swift",
    "//  This file was automatically generated and should not be edited.

import Apollo

public final class OtherQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query OtherQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}",
  ],
]
`;

exports[`successful codegen infers Flow target and writes types 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen infers JSON target and writes operations 1`] = `
"{
	\\"operations\\": [
		{
			\\"filePath\\": \\"queryOne.graphql\\",
			\\"operationName\\": \\"SimpleQuery\\",
			\\"operationType\\": \\"query\\",
			\\"rootType\\": \\"Query\\",
			\\"variables\\": [],
			\\"source\\": \\"query SimpleQuery {\\\\n  hello\\\\n}\\",
			\\"fields\\": [
				{
					\\"responseName\\": \\"hello\\",
					\\"fieldName\\": \\"hello\\",
					\\"type\\": \\"String!\\",
					\\"isConditional\\": false,
					\\"isDeprecated\\": false
				}
			],
			\\"fragmentSpreads\\": [],
			\\"inlineFragments\\": [],
			\\"fragmentsReferenced\\": [],
			\\"sourceWithFragments\\": \\"query SimpleQuery {\\\\n  hello\\\\n}\\",
			\\"operationId\\": \\"4de1e386589b0853a102a87a3f21ca25266116517e59c1e8be9442e2571f00bc\\"
		}
	],
	\\"fragments\\": [],
	\\"typesUsed\\": []
}"
`;

exports[`successful codegen infers Scala target and writes types 1`] = `
"//  This file was automatically generated and should not be edited.


object SimpleQueryQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query SimpleQuery {\\" +
    \\"  hello\\" +
    \\"}\\"
  val operation = com.apollographql.scalajs.gql(operationString)

  type Variables = Unit

  case class Data(hello: String) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }
}"
`;

exports[`successful codegen infers Swift target and writes types 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;

exports[`successful codegen infers Swift target and writes types when schema is a GraphQL file 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;

exports[`successful codegen infers Swift target and writes types when schema is a JS file 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;

exports[`successful codegen infers Swift target and writes types when schema is a TS file 1`] = `
"//  This file was automatically generated and should not be edited.

import Apollo

public final class SimpleQueryQuery: GraphQLQuery {
  public let operationDefinition =
    \\"query SimpleQuery {\\\\n  hello\\\\n}\\"

  public init() {
  }

  public struct Data: GraphQLSelectionSet {
    public static let possibleTypes = [\\"Query\\"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField(\\"hello\\", type: .nonNull(.scalar(String.self))),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(hello: String) {
      self.init(unsafeResultMap: [\\"__typename\\": \\"Query\\", \\"hello\\": hello])
    }

    public var hello: String {
      get {
        return resultMap[\\"hello\\"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: \\"hello\\")
      }
    }
  }
}"
`;

exports[`successful codegen infers TypeScript target and writes types 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen infers TypeScript target and writes types for query with client-side data 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery_complexLocalState {
  someData: string;
}

export interface SimpleQuery_serverSideField {
  serverData: string;
  addedLocalData: string;
}

export interface SimpleQuery_localServerModel {
  serverData: string;
}

export interface SimpleQuery {
  hello: string;
  localState: string;
  complexLocalState: SimpleQuery_complexLocalState;
  serverSideField: SimpleQuery_serverSideField;
  localServerModel: SimpleQuery_localServerModel;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen infers TypeScript target and writes types for query with client-side data with schema in a JS file 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery_complexLocalState {
  someData: string;
}

export interface SimpleQuery_serverSideField {
  serverData: string;
  addedLocalData: string;
}

export interface SimpleQuery_localServerModel {
  serverData: string;
}

export interface SimpleQuery {
  hello: string;
  localState: string;
  complexLocalState: SimpleQuery_complexLocalState;
  serverSideField: SimpleQuery_serverSideField;
  localServerModel: SimpleQuery_localServerModel;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen infers TypeScript target and writes types for query with client-side data with schema in a JS file from config 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery_complexLocalState {
  someData: string;
}

export interface SimpleQuery_serverSideField {
  serverData: string;
  addedLocalData: string;
}

export interface SimpleQuery_localServerModel {
  serverData: string;
}

export interface SimpleQuery {
  hello: string;
  localState: string;
  complexLocalState: SimpleQuery_complexLocalState;
  serverSideField: SimpleQuery_serverSideField;
  localServerModel: SimpleQuery_localServerModel;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen infers TypeScript target and writes types for query with only client-side data 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery_complexLocalState {
  someData: string;
}

export interface SimpleQuery {
  localState: string;
  complexLocalState: SimpleQuery_complexLocalState;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen writes Flow global types to a custom path when globalTypesFile is set 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

import { SomeEnum } from \\"./../../__foo__/bar.flow.js\\";

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
  someEnum: SomeEnum;
}
"
`;

exports[`successful codegen writes Flow global types to a custom path when globalTypesFile is set 2`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

export enum SomeEnum {
  bar = \\"bar\\",
  foo = \\"foo\\",
}

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen writes Flow types into a __generated__ directory next to sources when no output is set 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes Flow types next to sources when output is set to empty string 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes Flow types to a custom directory next to sources when output is set 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes TypeScript global types to a custom path when globalTypesFile is set 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

import { SomeEnum } from \\"./../../__foo__/bar\\";

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  someEnum: SomeEnum;
}
"
`;

exports[`successful codegen writes TypeScript global types to a custom path when globalTypesFile is set 2`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

export enum SomeEnum {
  bar = \\"bar\\",
  foo = \\"foo\\",
}

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen writes TypeScript types into a __generated__ directory next to sources when no output is set 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
}
"
`;

exports[`successful codegen writes TypeScript types into a __generated__ directory next to sources when no output is set 2`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen writes TypeScript types next to sources when output is set to empty string 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
}
"
`;

exports[`successful codegen writes TypeScript types next to sources when output is set to empty string 2`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen writes TypeScript types to a custom directory next to sources when output is set 1`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export interface SimpleQuery {
  hello: string;
}
"
`;

exports[`successful codegen writes TypeScript types to a custom directory next to sources when output is set 2`] = `
"/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//==============================================================
"
`;

exports[`successful codegen writes exact Flow types when the flag is set 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {|
  hello: string
|};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;

exports[`successful codegen writes read-only Flow types when the flag is set 1`] = `
"

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SimpleQuery
// ====================================================

export type SimpleQuery = {
  +hello: string
};

/* @flow */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

//==============================================================
// END Enums and Input Objects
//=============================================================="
`;
